README.md
A book about licensing R. 

rngwell19937/ChangeLog
2014-11-30  0.6-0  subdirectory orig_files moved to src
                   improving and correcting the documentation
2012-01-17  0.5-4  included NAMESPACE, update of the documentation
2009-08-16  0.5-3  extended DESCRIPTION file
                   corrected ?set.vector.seed
                   code cleaning
2009-04-16  0.5-2  corrected error message in set.initialization() thanks to Jakub Dvorak
                   extended documentation
2009-03-12  0.5-1  add test of the current RNGkind() to the validation function
2009-03-05  0.5-0  initial CRAN submission


rngwell19937/src/original_files/combmrg2.c

 #define norm 2.328306549295728e-10
 #define m1   4294967087.0
 #define m2   4294944443.0
 #define a12     1403580.0
 #define a13n     810728.0
 #define a21      527612.0
 #define a23n    1370589.0

 double  s10, s11, s12, s20, s21, s22;

 double MRG32k3a ()
    {
    long   k;
    double p1, p2;
    /* Component 1 */
    p1 = a12 * s11 - a13n * s10;
    k = p1 / m1;   p1 -= k * m1;   if (p1 < 0.0) p1 += m1;
    s10 = s11;   s11 = s12;   s12 = p1;
    /* Component 2 */
    p2 = a21 * s22 - a23n * s20;
    k  = p2 / m2;  p2 -= k * m2;   if (p2 < 0.0) p2 += m2;
    s20 = s21;   s21 = s22;   s22 = p2;
    /* Combination */
    if (p1 <= p2) return ((p1 - p2 + m1) * norm);
    else return ((p1 - p2) * norm);
    }

######################

 double  s10, s11, s12, s13, s14, s20, s21, s22, s23, s24;

 #define norm 2.3283163396834613e-10
 #define m1   4294949027.0
 #define m2   4294934327.0
 #define a12     1154721.0
 #define a14     1739991.0
 #define a15n    1108499.0
 #define a21     1776413.0
 #define a23      865203.0
 #define a25n    1641052.0

double MRG32k5a ()
    {
    long   k;
    double p1, p2;
    /* Component 1 */
    p1 = a12 * s13 - a15n * s10;
    if (p1 > 0.0) p1 -= a14 * m1;
    p1 += a14 * s11;   k = p1 / m1;    p1 -= k * m1;
    if (p1 < 0.0) p1 += m1;
    s10 = s11;   s11 = s12;   s12 = s13;  s13 = s14;  s14 = p1;
    /* Component 2 */
    p2 = a21 * s24 - a25n * s20;
    if (p2 > 0.0) p2 -= a23 * m2;
    p2 += a23 * s22;    k  = p2 / m2;   p2 -= k * m2;
    if (p2 < 0.0) p2 += m2;
    s20 = s21;   s21 = s22;   s22 = s23;  s23 = s24;  s24 = p2;
    /* Combination */
    if (p1 <= p2)  return ((p1 - p2 + m1) * norm);
    else  return ((p1 - p2) * norm);
    }

######################

 #define norm  1.0842021724855052e-19
 #define m1    9223372036854769163
 #define m2    9223372036854754679
 #define a12   1754669720
 #define q12   5256471877
 #define r12   251304723
 #define a13n  3182104042
 #define q13   2898513661
 #define r13   394451401
 #define a21   31387477935
 #define q21   293855150
 #define r21   143639429
 #define a23n  6199136374
 #define q23   1487847900
 #define r23   985240079

 long long s10, s11, s12, s20, s21, s22;

 double MRG63k3a ()
    {
    long long h, p12, p13, p21, p23;
    /* Component 1 */
    h = s10 / q13;   p13 = a13n * (s10 - h * q13) - h * r13;
    h = s11 / q12;   p12 = a12  * (s11 - h * q12) - h * r12;
    if (p13 < 0) p13 += m1;    
    if (p12 < 0) p12 += m1 - p13;  else p12 -= p13;
    if (p12 < 0) p12 += m1;    
    s10 = s11;  s11 = s12;  s12 = p12;
    /* Component 2 */
    h = s20 / q23;   p23 = a23n * (s20 - h * q23) - h * r23;
    h = s22 / q21;   p21 = a21  * (s22 - h * q21) - h * r21;
    if (p23 < 0) p23 += m2;
    if (p21 < 0) p21 += m2 - p23;  else p21 -= p23;
    if (p21 < 0) p21 += m2;    
    s20 = s21;  s21 = s22;  s22 = p21;
    /* Combination */
    if (p12 > p21) return ((p12 - p21) * norm);
    else return ((p12 - p21 + m1) * norm);
    }



rngwell19937/src/original_files/WELL19937a.c
/* ***************************************************************************** */
/* Copyright:      Francois Panneton and Pierre L'Ecuyer, University of Montreal */
/*                 Makoto Matsumoto, Hiroshima University                        */
/* Notice:         This code can be used freely for personal, academic,          */
/*                 or non-commercial purposes. For commercial purposes,          */
/*                 please contact P. L'Ecuyer at: lecuyer@iro.UMontreal.ca       */
/* ***************************************************************************** */

#define W 32
#define R 624
#define P 31
#define MASKU (0xffffffffU>>(W-P))
#define MASKL (~MASKU)
#define M1 70
#define M2 179
#define M3 449

#define MAT0POS(t,v) (v^(v>>t))
#define MAT0NEG(t,v) (v^(v<<(-(t))))
#define MAT1(v) v
#define MAT3POS(t,v) (v>>t)

/* To obtain the WELL19937c, uncomment the following line */
/* #define TEMPERING                                      */
#define TEMPERB 0xe46e1700U
#define TEMPERC 0x9b868000U

#define V0            STATE[state_i]
#define VM1Over       STATE[state_i+M1-R]
#define VM1           STATE[state_i+M1]
#define VM2Over       STATE[state_i+M2-R]
#define VM2           STATE[state_i+M2]
#define VM3Over       STATE[state_i+M3-R]
#define VM3           STATE[state_i+M3]
#define VRm1          STATE[state_i-1]
#define VRm1Under     STATE[state_i+R-1]
#define VRm2          STATE[state_i-2]
#define VRm2Under     STATE[state_i+R-2]

#define newV0         STATE[state_i-1]
#define newV0Under    STATE[state_i-1+R]
#define newV1         STATE[state_i]
#define newVRm1       STATE[state_i-2]
#define newVRm1Under  STATE[state_i-2+R]

#define FACT 2.32830643653869628906e-10

static int state_i = 0;
static unsigned int STATE[R];
static unsigned int z0, z1, z2;
static double case_1 (void);
static double case_2 (void);
static double case_3 (void);
static double case_4 (void);
static double case_5 (void);
static double case_6 (void);
       double (*WELLRNG19937a) (void);

static unsigned int y;

void InitWELLRNG19937a (unsigned int *init){
   int j;
   state_i = 0;
   WELLRNG19937a = case_1;
   for (j = 0; j < R; j++)
     STATE[j] = init[j];
}

double case_1 (void){
   // state_i == 0
   z0 = (VRm1Under & MASKL) | (VRm2Under & MASKU);
   z1 = MAT0NEG (-25, V0) ^ MAT0POS (27, VM1);
   z2 = MAT3POS (9, VM2) ^ MAT0POS (1, VM3);
   newV1      = z1 ^ z2;
   newV0Under = MAT1 (z0) ^ MAT0NEG (-9, z1) ^ MAT0NEG (-21, z2) ^ MAT0POS (21, newV1);
   state_i = R - 1;
   WELLRNG19937a = case_3;
#ifdef TEMPERING
   y = STATE[state_i] ^ ((STATE[state_i] << 7) & TEMPERB);
   y =              y ^ ((             y << 15) & TEMPERC);
   return ((double) y * FACT);
#else
   return ((double) STATE[state_i] * FACT);
#endif
}

static double case_2 (void){
   // state_i == 1
   z0 = (VRm1 & MASKL) | (VRm2Under & MASKU);
   z1 = MAT0NEG (-25, V0) ^ MAT0POS (27, VM1);
   z2 = MAT3POS (9, VM2) ^ MAT0POS (1, VM3);
   newV1 = z1 ^ z2;
   newV0 = MAT1 (z0) ^ MAT0NEG (-9, z1) ^ MAT0NEG (-21, z2) ^ MAT0POS (21, newV1);
   state_i = 0;
   WELLRNG19937a = case_1;
#ifdef TEMPERING
   y = STATE[state_i] ^ ((STATE[state_i] << 7) & TEMPERB);
   y =              y ^ ((             y << 15) & TEMPERC);
   return ((double) y * FACT);
#else
   return ((double) STATE[state_i] * FACT);
#endif
}

static double case_3 (void){
   // state_i+M1 >= R
   z0 = (VRm1 & MASKL) | (VRm2 & MASKU);
   z1 = MAT0NEG (-25, V0) ^ MAT0POS (27, VM1Over);
   z2 = MAT3POS (9, VM2Over) ^ MAT0POS (1, VM3Over);
   newV1 = z1 ^ z2;
   newV0 = MAT1 (z0) ^ MAT0NEG (-9, z1) ^ MAT0NEG (-21, z2) ^ MAT0POS (21, newV1);
   state_i--;
   if (state_i + M1 < R)
      WELLRNG19937a = case_5;
#ifdef TEMPERING
   y = STATE[state_i] ^ ((STATE[state_i] << 7) & TEMPERB);
   y =              y ^ ((             y << 15) & TEMPERC);
   return ((double) y * FACT);
#else
   return ((double) STATE[state_i] * FACT);
#endif
}

static double case_4 (void){
   // state_i+M3 >= R
   z0 = (VRm1 & MASKL) | (VRm2 & MASKU);
   z1 = MAT0NEG (-25, V0) ^ MAT0POS (27, VM1);
   z2 = MAT3POS (9, VM2) ^ MAT0POS (1, VM3Over);
   newV1 = z1 ^ z2;
   newV0 = MAT1 (z0) ^ MAT0NEG (-9, z1) ^ MAT0NEG (-21, z2) ^ MAT0POS (21, newV1);
   state_i--;
   if (state_i + M3 < R)
      WELLRNG19937a = case_6;
#ifdef TEMPERING
   y = STATE[state_i] ^ ((STATE[state_i] << 7) & TEMPERB);
   y =              y ^ ((             y << 15) & TEMPERC);
   return ((double) y * FACT);
#else
   return ((double) STATE[state_i] * FACT);
#endif
}

static double case_5 (void){
   // state_i+M2 >= R
   z0 = (VRm1 & MASKL) | (VRm2 & MASKU);
   z1 = MAT0NEG (-25, V0) ^ MAT0POS (27, VM1);
   z2 = MAT3POS (9, VM2Over) ^ MAT0POS (1, VM3Over);
   newV1 = z1 ^ z2;
   newV0 = MAT1 (z0) ^ MAT0NEG (-9, z1) ^ MAT0NEG (-21, z2) ^ MAT0POS (21, newV1);
   state_i--;
   if (state_i + M2 < R)
      WELLRNG19937a = case_4;
#ifdef TEMPERING
   y = STATE[state_i] ^ ((STATE[state_i] << 7) & TEMPERB);
   y =              y ^ ((             y << 15) & TEMPERC);
   return ((double) y * FACT);
#else
   return ((double) STATE[state_i] * FACT);
#endif
}

static double case_6 (void){
   // 2 <= state_i <= (R - M3 - 1)
   z0 = (VRm1 & MASKL) | (VRm2 & MASKU);
   z1 = MAT0NEG (-25, V0) ^ MAT0POS (27, VM1);
   z2 = MAT3POS (9, VM2) ^ MAT0POS (1, VM3);
   newV1 = z1 ^ z2;
   newV0 = MAT1 (z0) ^ MAT0NEG (-9, z1) ^ MAT0NEG (-21, z2) ^ MAT0POS (21, newV1);
   state_i--;
   if (state_i == 1)
      WELLRNG19937a = case_2;
#ifdef TEMPERING
   y = STATE[state_i] ^ ((STATE[state_i] << 7) & TEMPERB);
   y =              y ^ ((             y << 15) & TEMPERC);
   return ((double) y * FACT);
#else
   return ((double) STATE[state_i] * FACT);
#endif
}



rngwell19937/src/interface.c
#include "R.h"
#include <R_ext/Random.h>

#define m1   4294949027.0
#define m2   4294934327.0
extern double  s10, s11, s12, s13, s14, s20, s21, s22, s23, s24;
unsigned int MRG32k5a ();

void InitWELLRNG19937a ();
extern unsigned int (*WELLRNG19937a) (void);
void select_case();

#define R19937a   624
static unsigned int random_seed[1+R19937a]; // for interface to R
static int seedlength[1] = {1+R19937a};     // for interface to R
unsigned int *rstate_i = random_seed;  // used in WELLRNG19937a()
unsigned int *STATE = random_seed+1;   // used in WELLRNG19937a()
unsigned int check_state_i;  // detect user modification of .Random.seed

#define bin32m 2.3283064365386962891e-10 // 2^(-32)
#define bin33m 1.1641532182693481445e-10 // 2^(-33)
#define bin53m 1.1102230246251565404e-16 // 2^(-53)
#define bin54m 5.5511151231257827021e-17 // 2^(-54)

double x;

void get_rand_32(void)
{
	x = WELLRNG19937a() * bin32m;
	if (x == 0) {
		x = bin33m;
	}
}

void get_rand_53(void)
{
	x = (double) (WELLRNG19937a() >> 6);
	x = (134217728.0 * x + (double) (WELLRNG19937a() >> 5)) * bin53m;
	if (x == 0) {
		x = bin54m;
	}
}

static void (*get_rand) (void) = get_rand_53;

void set_resolution(int *r)
{
	if (*r) {
		get_rand = get_rand_53;
	} else {
		get_rand = get_rand_32;
	}
}

double *user_unif_rand(void)
{
	if (random_seed[0] != check_state_i) {
		select_case();
	}
	get_rand();
	check_state_i = random_seed[0];
	return(&x);
}

void init_mrg32k5a(int n, unsigned int *seed)
{
	int i,j;
	unsigned int l16, h16;
	s10 = 12345.0;
	s11 = 12345.0;
	s12 = 12345.0;
	s13 = 12345.0;
	s14 = 12345.0;
	s20 = 12345.0;
	s21 = 12345.0;
	s22 = 12345.0;
	s23 = 12345.0;
	s24 = 12345.0;
	for (i=0; i < n; i++) {
		l16 = seed[i] & 0x0000ffff;
		h16 = seed[i] >> 16;
		s14 += (double) h16 + 1.0;
		s24 += (double) h16 + 1.0;
		if (s14 >= m1) { s14 -= m1; }
		if (s24 >= m2) { s24 -= m2; }
		for (j=0; j<5; j++) MRG32k5a();
		s14 += (double) l16 + 1.0;
		s24 += (double) l16 + 1.0;
		if (s14 >= m1) { s14 -= m1; }
		if (s24 >= m2) { s24 -= m2; }
		for (j=0; j<5; j++) MRG32k5a();
	}
	for (i = 0; i < R19937a; i++) {
		STATE[i] = (MRG32k5a() & 0x0000ffff);
		STATE[i] = (STATE[i] << 16) | (MRG32k5a() & 0x0000ffff);
	}
	InitWELLRNG19937a();
}

void init_vector_mrg32k5a(int *n, double *seed, unsigned int *iseed, unsigned int *state)
{
	int i;
	for (i=0; i < *n; i++) {
		iseed[i] = (unsigned int) seed[i];
	}
	init_mrg32k5a(*n, iseed);
	for (i=0; i < 625; i++) {
		state[i] = random_seed[i];
	}
}

void user_unif_init_mrg32k5a(unsigned int seed)
{
	init_mrg32k5a(1, &seed);
}

void user_unif_init_sfmt(unsigned int seed)
{
	int i;
	STATE[0] = seed;
	for (i = 1; i < R19937a; i++) 
		STATE[i] = 1812433253UL * ( STATE[i - 1] ^ ( STATE[i - 1] >> 30 ) ) + i;
	InitWELLRNG19937a();
}

static void (*user_unif_init_selected) (unsigned int seed) = user_unif_init_mrg32k5a;

void set_initialization(int *initialization)
{
	if (*initialization) {
		user_unif_init_selected = user_unif_init_mrg32k5a;
	} else {
		user_unif_init_selected = user_unif_init_sfmt;
	}
}

void user_unif_init(unsigned int seed)
{
	seed = 3602842457U * seed + 105890386U; // undo initial scrambling
	user_unif_init_selected(seed);
	check_state_i = random_seed[0];
}

int *user_unif_nseed(void)
{
	return (seedlength);
}

int *user_unif_seedloc(void)
{
	return ((int *) random_seed);
}

